insert next-gen permission paths alongside currently active permission paths + classification

escherize opened the pull request:
resolves #27300

Most importantly this PR adds the function `->v2-path` which is called on v1 permission path strings just before they are written into the app db. `->v2-path` returns a vector of v2 compliant paths, which are then inserted into the DB.

Another thing it does is refactor `u.regex/rx` from a macro to a function, and the things it operates on from lists of symbols to vectors of keywords. Seems like a small thing but it is a lot easier to use. I considered swapping over to *regal*, which is similar, but this is a pretty sensitive and mistakes are easy in this area, so didn't want to do the port at this time.

Notable tests include:

1. path classification tests: https://github.com/metabase/metabase/pull/27911/files#diff-1ac2b6f3424a7074e53184da0e863b7d6b6bc5ae1e466a30566dc30853fdbf57R861-R894
2. move idempotency test: https://github.com/metabase/metabase/pull/27911/files#diff-1ac2b6f3424a7074e53184da0e863b7d6b6bc5ae1e466a30566dc30853fdbf57R896
3. quickcheck `classify-path` `classify-data-path` and `->v2-path`, which costs about 10 loc! : https://github.com/metabase/metabase/pull/27911/files#diff-1ac2b6f3424a7074e53184da0e863b7d6b6bc5ae1e466a30566dc30853fdbf57R896

FAQ:

What is a v1 or v2 path?
They're defined by this table. https://www.notion.so/metabase/Permissions-Refactor-Design-Doc-18ff5e6be32f4a52b9422bd7f4237ca7#5603afe084a7435ca7dc928fc94d4bda

escherize committed:
implement move, which returns v2 paths

- TODO: insert these into the db

(move v1-path) => [v2 paths]

escherize committed:
cleanup + add some schemas

escherize committed:
generative tests 4 permission path classification

escherize committed:
whitespace lint

escherize committed:
detect data, query, and paths for v2

escherize committed:
calling move on v2 paths is a no-op

escherize committed:
differentiate between v1 and v2 permissions

quickchecking for move, classify-path, and classify-data-path

escherize committed:
fix tests + add idempotency test

escherize committed:
add tests for classification of permission paths

- rename move to ->v2-path
- move some fxns around
- ascii art in test

escherize committed:
making the legos line up

- need to insert both v1 and v2 versions of paths (of course)
- valid-path? has to allow v2 paths to be inserted

escherize committed:
replace mu/with-api-error-message

escherize committed:
linter + code quality fixes

escherize committed:
privatize rx->kind

escherize committed:
Merge branch 'master' into bcm-write-v2-perm-paths

escherize committed:
remove some changes that should be 4 anotherbranch

escherize committed:
Merge branch 'master' into bcm-write-v2-perm-paths

escherize committed:
Merge branch 'bcm-write-v2-perm-paths' of github.com:metabase/metabase into bcm-write-v2-perm-paths

escherize commented on src/metabase/models/permissions.clj:
```diff
@@ -221,93 +221,169 @@
-(def ^:private path-regex
+  (u.regex/rx [:or #"[^\\/]" #"\\/" #"\\\\"]))
+
+(def ^:private data-rx->data-kind
+  {      #"db/\d+/"                                                                  :dk/db
+   [:and #"db/\d+/" "native" "/"]                                                    :dk/db-native
+   [:and #"db/\d+/" "schema" "/"]                                                    :dk/db-schema
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/"]                                  :dk/db-schema-name
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/table/\\d+/"]                       :dk/db-schema-name-and-table
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/table/\\d+/" "read/"]               :dk/db-schema-name-table-and-read
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/table/\\d+/" "query/"]              :dk/db-schema-name-table-and-query
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/table/\\d+/" "query/" "segmented/"] :dk/db-schema-name-table-and-segmented})
```
this could be more performant, yes. But IMO the clarity here is better, at least until we are finished upgrading perm-path formats.

escherize committed:
revert ns

escherize committed:
Merge branch 'master' into bcm-write-v2-perm-paths

escherize committed:
Merge branch 'master' into bcm-write-v2-perm-paths

escherize committed:
Merge branch 'master' into bcm-write-v2-perm-paths

metamben commented on src/metabase/models/permissions.clj:
```diff
@@ -853,6 +927,41 @@
+(let [delete (fn delete [s to-delete] (str/replace s to-delete ""))
+      data-query-split (fn data-query-split [path]
+                         [(str "/data" path) (str "/query" path)])]
+  (def ^:private data-kind->rewrite-fn
+    "lookup table to generate v2 query + data permission from a v1 data permission."
+    {:dk/db                                 data-query-split
+     :dk/db-native                          (fn [path] (data-query-split (delete path "native/")))
+     :dk/db-schema                          (fn [path] [(str "/data" (delete path "schema/"))
+                                                        (str "/query" path)])
+     :dk/db-schema-name                     data-query-split
+     :dk/db-schema-name-and-table           data-query-split
+     :dk/db-schema-name-table-and-read      (constantly [])
+     :dk/db-schema-name-table-and-query     (fn [path] (data-query-split (delete path "query/")))
+     :dk/db-schema-name-table-and-segmented (fn [path] (data-query-split (delete path "query/segmented/")))}))
```
Is this a homegrown multi-fn?

escherize replied:
It's a map of keywords to functions. Could you flesh out your comment some more? I can't tell what you mean.

metamben commented on src/metabase/models/permissions.clj:
```diff
@@ -853,6 +927,41 @@
+      (= kind :admin) ["/"]
+      (= kind :block) []
+
+      ;; explicitly, move is idempotent mapcatting move over a sequence of v1 paths multiple times will result in the same value.
+      ;; so these are no-ops:
+      (= kind :data-v2) [path]
+      (= kind :query-v2) [path]
```
Should this comment explain why these cases are explicitly listed?

escherize replied:
It should. I'll make it more clear.

metamben commented on src/metabase/models/permissions.clj:
```diff
@@ -853,6 +927,41 @@
+  [path :- [:or [:re path-regex-v1] [:re path-regex-v2]]]
+  ;; See: https://www.notion.so/metabase/Permissions-Refactor-Design-Doc-18ff5e6be32f4a52b9422bd7f4237ca7#5603afe084a7435ca7dc928fc94d4bda
+  (let [kind (classify-path path)]
+    (cond
```
Could be `(case kind`?

metamben reviewed (approved):
Works for me, but I suggest getting another opinion about replacing the regex macros with functions.

escherize commented on src/metabase/models/permissions.clj:
```diff
@@ -221,93 +221,169 @@
+   [:and #"db/\d+/" "native" "/"]                                                    :dk/db-native
+   [:and #"db/\d+/" "schema" "/"]                                                    :dk/db-schema
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/"]                                  :dk/db-schema-name
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/table/\\d+/"]                       :dk/db-schema-name-and-table
```
These could be any number of things. They get converted into a regex.

I don't see what you're getting at, please be more direct.

noahmoss replied:
I think this is in line with the intended use of `u.regex/rx`

noahmoss commented on src/metabase/models/permissions.clj:
```diff
@@ -853,6 +927,41 @@
   (delete-related-permissions! group-or-id (apply (partial feature-perms-path :download :full) path-components))
   (delete-related-permissions! group-or-id (apply (partial feature-perms-path :download :limited) path-components)))
 
+(let [delete (fn delete [s to-delete] (str/replace s to-delete ""))
```
FWIW I usually prefer `let ... fn` to `letfn` since it makes it easier to capture the function in a var and call it directly in the REPL when debugging

noahmoss commented on src/metabase/models/permissions.clj:
```diff
@@ -1146,11 +1256,16 @@
   (doseq [[db-id new-db-perms] new-group-perms
           [perm-type new-perms] new-db-perms]
     (case perm-type
+      ;; TODO  use query
```
This is a TODO for the next PR right?

noahmoss commented on test/metabase/models/permissions_test.clj:
```diff
@@ -856,3 +857,115 @@
+    (is (= v2-paths (mapcat #'perms/->v2-path v1-paths)))
+    (is (= v2-paths (mapcat #'perms/->v2-path v2-paths)))
+    (let [w (partial mapcat #'perms/->v2-path)]
+      (is (= v2-paths (->
+                                    v1-paths
+                          w w                       w w;
+                          w w                       w w;
+                          w w w w w w w w w w w w w w w;
+                          w w w w w w w w w w w w w w w;
+                          w w w                   w w w;
+                          w w      w         w      w w
+                          w w     w w       w w     w w
+                          w w           w           w w
+                          w w           w           w w
+                          w w w w w w w w w w w w w w w;
+                              w w w w w w w w w w w;;;;
+                              w w w w w w w w w w w;;
+                                  w w w w w w w;
+                                      w   w;
+                                      w   w
+                                    w w   w w;
+                                    ) ) ) ) ) )
+
```
🦉

noahmoss replied:
(nit: doesn't really matter since the formatting is wild here anyway but there's an extra newline after this test)

noahmoss reviewed (approved):
Awesome, looks good to me. I like the way you've broken down the regex.

escherize committed:
delete v2 permissions in

- they aren't handled by the perm graph parser, so they don't get propagated into "old graph", so the diff between old and new indicates that they need to be rewritten.

escherize committed:
Merge branch 'master' into bcm-write-v2-perm-paths

escherize committed:
Merge branch 'bcm-write-v2-perm-paths' of github.com:metabase/metabase into bcm-write-v2-perm-paths

escherize committed:
only delete v2 paths for the current group_id

- reorder declarations in models.permissions

escherize committed:
remove extra line break

escherize committed:
Update src/metabase/models/permissions.clj

fix typo

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>
